irods_audit_elk_stack/not-logstash/not-logstash.py

```diff
@@ -14,52 +18,128 @@
 # I think that we have to do this at all means we're constructing our json improperly.
 # Ugh.
 convert_int = [
-  'int',
-  'int__2',
-  'int__3',
-  'file_size'
+	'int',
+	'int__2',
+	'int__3',
+	'file_size'
 ]
 
+def get_argparser():
+	default_mut = Workarounds.TYPING
+	parser = argparse.ArgumentParser(
+		description='AMQP 1.0 fluent Logstash stand-in daemon',
+		allow_abbrev=False
+	)
+	parser.add_argument(
+		'--workarounds', metavar='MUTATOR',
+		type=Workarounds,
+		action=FlagAction, dest='workarounds', nargs='+',
+		default=default_mut,
+		help=f'Specify which workarounds to enable (available: {~Workarounds.NONE}; default: {default_mut})'
+	)
+	return parser
+
+class FlagArg(Flag):
+	@classmethod
+	def _from_arg_str(cls, s: str) -> Flag:
+		value = cls(0)
+		name_map = {str(f): f for f in cls}
+		names = [n for n in s.split(',') if n]
+		for name in names:
+			try:
+				value = value | name_map[name]
+			except KeyError:
+				raise argparse.ArgumentTypeError(
+					f"{name!r} is not a valid {cls.__name__}")
+		return value
+
+	def __str__(self):
+		if self.name is not None:
+			return self.name.lower()
+		members, other = _decompose(self.__class__, self.value)
+		return ','.join([str(m.name or m.value).lower() for m in members])
+
+class FlagAction(argparse._StoreAction):
+	def __init__(self, *args, type=None, **kwargs):
+		super(FlagAction, self).__init__(*args, type=type._from_arg_str, **kwargs)
+		self.flag_type = type
+
+	def __call__(self, parser, namespace, values, *args, **kwargs):
+		value = self.flag_type(0)
+		for v in values:
+			value = value | v
+		super(FlagAction, self).__call__(parser, namespace, value, *args, **kwargs)
+
+class Workarounds(FlagArg, Flag):
+	NONE = 0
+	TOKENS = auto()
+	TYPING = auto()
+	TIMESTAMP = auto()
+
 class ELKReader(MessagingHandler):
-  def __init__(self, server, address):
-    super(ELKReader, self).__init__()
-    self.server = server
-    self.address = address
-    self.id_ctr = 0
-    self.id_ctr_lock = Lock()
-    self.es = Elasticsearch(hosts=[{'host':'localhost', 'port':9200}])
+	def __init__(self, server, address, workarounds=Workarounds.NONE):
+		self.logger = logging.getLogger('ELKReader')
+		super(ELKReader, self).__init__()
+		self.workaround_tokens = self._w_tokens if Workarounds.TOKENS in workarounds else self._w_noop
+		self.workaround_typing = self._w_typing if Workarounds.TYPING in workarounds else self._w_noop
+		self.workaround_ts = self._w_ts if Workarounds.TIMESTAMP in workarounds else self._w_noop
+		self.server = server
+		self.address = address
+		self.id_ctr = 0
+		self.id_ctr_lock = Lock()
+		self.es = Elasticsearch(hosts=[{'host':'localhost', 'port':9200}])
+
+	def on_start(self, event):
+		conn = event.container.connect(self.server)
+		event.container.create_receiver(conn, self.address)
+
+	def _w_noop(self, *args, **kwargs):
```

Comment:
Are these function names special in any way (e.g. inheritance)?

Reply:
Not syntactically. I'll break it down:
1. `_`: Starting symbol names with a single underscore is standard practice for indicating that they are not meant to be referenced outside of the module or the class/class' children.
2. `w`: w for workaround
3. `_`: spacer
4. `noop`: name of workaround. This one doesn't do anything, so `noop`.

I did the workarounds in this manner so that they can all be switched on or off at the start, rather than checking every workaround for every message every time.

Reply:
I felt that is what was happening (especially seeing the `__init__` impl).

Thanks for the explanation. This is good.